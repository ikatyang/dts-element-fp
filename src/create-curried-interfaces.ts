import * as dts from 'dts-element';
import * as R from 'ramda';
import {
  get_interface_generic_name_default,
  get_interface_generic_return_name_default,
  get_interface_name_default,
  get_interface_parameter_name_default,
  get_interface_parameter_placeholder_name_default,
  get_placeholder_type_default,
  placeholder_default,
  selectable_default,
} from './utils/constants';
import {
  create_selectable_signatures,
  CreateSelectableSignaturesOptions,
} from './utils/create-selectable-signatures';
import { push_signatures } from './utils/push-signatures';
import { sort_signatures } from './utils/sort-signatures';

export interface CreateCurriedInterfacesOptions
  extends CreateSelectableSignaturesOptions {
  selectable?: boolean;
  placeholder?: boolean;
  get_placeholder_type?: () => string;
  get_interface_name?: (index: number) => string;
  get_generic_name?: (index: number) => string;
  get_generic_return_name?: () => string;
  get_parameter_name?: (index: number) => string;
  get_parameter_placeholder_name?: (index: number) => string;
}

export function create_curried_interfaces(
  max_curry_level: number,
  // istanbul ignore next
  options: CreateCurriedInterfacesOptions = {},
): dts.IInterfaceDeclaration[] {
  // istanbul ignore next
  const {
    selectable = selectable_default,
    placeholder = placeholder_default,
    get_placeholder_type = get_placeholder_type_default,
    get_interface_name = get_interface_name_default,
    get_generic_name = get_interface_generic_name_default,
    get_generic_return_name = get_interface_generic_return_name_default,
    get_parameter_name = get_interface_parameter_name_default,
    get_parameter_placeholder_name = get_interface_parameter_placeholder_name_default,
    get_selectable_kind_name,
    get_selectable_selector_name,
  } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const generics = R.range(0, max_curry_level).map((_, index) =>
    dts.create_generic_declaration({ name: get_generic_name(index) }),
  );

  const generic_return = dts.create_generic_declaration({
    name: get_generic_return_name(),
  });

  const parameters = generics.map((generic, index) =>
    dts.create_parameter_declaration({
      name: get_parameter_name(index),
      type: dts.create_general_type({ name: generic.name }),
    }),
  );

  const placeholders = parameters.map((_, index) =>
    dts.create_parameter_declaration({
      name: get_parameter_placeholder_name(index),
      type: dts.create_general_type({ name: get_placeholder_type() }),
    }),
  );

  const interfaces = generics.map((_, index) =>
    dts.create_interface_declaration({
      name: get_interface_name(index),
      generics: [...generics.slice(0, index + 1), generic_return],
      type: dts.create_object_type({ members: [] }),
    }),
  );

  interfaces.forEach((the_interface, index) => {
    push_signatures(
      is_placeholder,
      parameters.slice(0, index + 1),
      placeholder ? placeholders.slice(0, index + 1) : [],
      (used_parameters, return_parameters) => {
        the_interface.type!.members!.push(
          dts.create_object_member({
            owned: dts.create_function_declaration({
              name: undefined,
              type: dts.create_function_type({
                parameters: used_parameters,
                return:
                  return_parameters.length === 0
                    ? dts.create_general_type({ name: generic_return.name })
                    : dts.create_general_type({
                        name: interfaces[return_parameters.length - 1].name,
                        generics: [
                          ...return_parameters.map(
                            parameter => parameter.type!,
                          ),
                          dts.create_general_type({
                            name: generic_return.name,
                          }),
                        ],
                      }),
              }),
            }),
          }),
        );
      },
    );

    const object_type = interfaces[index].type!;
    object_type.members = sort_signatures(
      object_type.members as dts.IObjectMember[],
      { is_placeholder },
    );

    if (!selectable || object_type.members.length <= 1) {
      return;
    }

    object_type.members.splice(
      -1,
      0,
      ...create_selectable_signatures(
        object_type.members as dts.IObjectMember[],
        {
          get_selectable_kind_name,
          get_selectable_selector_name,
          is_placeholder,
        },
      ).reverse(),
    );
  });

  return interfaces;
}
