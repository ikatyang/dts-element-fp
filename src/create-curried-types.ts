import * as dts from 'dts-element';
import * as R from 'ramda';
import {
  get_function_parameter_placeholder_name_default,
  get_function_type_name_default,
  get_placeholder_type_default,
  inline_return_type_default,
  placeholder_default,
  selectable_default,
} from './utils/constants';
import { create_masks } from './utils/create-masks';
import {
  create_selectable_signatures,
  CreateSelectableSignaturesOptions,
} from './utils/create-selectable-signatures';
import { get_generics_dependencies } from './utils/get-generics-dependencies';
import { has } from './utils/has';
import { push_signatures } from './utils/push-signatures';

export interface CreateCurriedDeclarationsOptions
  extends CreateSelectableSignaturesOptions {
  selectable?: boolean;
  placeholder?: boolean;
  get_placeholder_type?: () => string;
  get_function_type_name?: (name: string, mask: string) => string;
  get_function_parameter_placeholder_name?: (name: string) => string;
  inline_return_type?: boolean;
}

/**
 * @hidden
 */
export function create_curried_types(
  name: string,
  type: dts.IFunctionType,
  // istanbul ignore next
  options: CreateCurriedDeclarationsOptions = {},
): dts.ITypeDeclaration[] {
  // istanbul ignore next
  const {
    selectable = selectable_default,
    placeholder = placeholder_default,
    get_placeholder_type = get_placeholder_type_default,
    get_function_type_name = get_function_type_name_default,
    get_function_parameter_placeholder_name = get_function_parameter_placeholder_name_default,
    get_selectable_kind_name,
    get_selectable_selector_name,
    inline_return_type = inline_return_type_default,
  } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const { generics = [], parameters = [], return: return_type } = type;

  const type_predicate_parameter = dts.is_type_predicate(return_type)
    ? parameters.find(
        parameter => parameter.name === (return_type.parameter as string),
      )!
    : null;

  const placeholders = parameters.map(parameter =>
    dts.create_parameter_declaration({
      name: get_function_parameter_placeholder_name(parameter.name),
      type: dts.create_general_type({ name: get_placeholder_type() }),
    }),
  );

  const generics_dependencies = get_generics_dependencies(generics);

  const sort_generics = (the_generics: dts.IGenericDeclaration[]) =>
    the_generics
      .slice()
      .sort((a, b) => generics.indexOf(a) - generics.indexOf(b));

  const add_generic_dependencies = (
    the_generics: dts.IGenericDeclaration[],
  ) => {
    const added: dts.IGenericDeclaration[] = [];

    the_generics.forEach(generic => {
      added.push(...generics_dependencies[generics.indexOf(generic)]);
    });

    return sort_generics(R.uniq(added));
  };

  const parameters_generics = parameters.map(parameter =>
    add_generic_dependencies(
      generics.filter(generic =>
        has(parameter, {
          kind: dts.ElementKind.GeneralType,
          name: generic.name,
        }),
      ),
    ),
  );

  const return_generics = add_generic_dependencies(
    generics.filter(generic =>
      has(return_type, {
        kind: dts.ElementKind.GeneralType,
        name: generic.name,
      }),
    ),
  );

  const target_types = R.repeat(0, 2 ** parameters.length).map(
    (_, index, array) =>
      index === array.length - 1
        ? dts.is_type_predicate(return_type) ? dts.boolean_type : return_type
        : dts.create_object_type({ members: [] }),
  );

  const masks = create_masks(parameters.length);
  const reverse_masks = masks.reduce<{ [mask: string]: number }>(
    (current_reverse_masks, mask, index) => ({
      ...current_reverse_masks,
      [mask]: index,
    }),
    {},
  );

  const type_declarations = target_types.map((target_type, index) => {
    const type_generics = sort_generics(
      index === target_types.length - 1
        ? return_generics
        : parameters_generics
            .filter((_, generics_index) => masks[index][generics_index] === '1')
            .reduce(
              (current_type_generics, filtered_parameter_generics) =>
                R.uniq(
                  current_type_generics.concat(filtered_parameter_generics),
                ),
              [],
            ),
    );
    return dts.create_type_declaration({
      name: get_function_type_name(name, masks[index]),
      generics: type_generics,
      type: target_type,
    });
  });

  type_declarations.slice(0, -1).forEach((type_declaration, index) => {
    const members = (type_declaration.type as dts.IObjectType)
      .members as dts.IObjectMember[];
    push_signatures(
      is_placeholder,
      parameters.filter((_, param_index) => masks[index][param_index] === '0'),
      placeholder
        ? placeholders.filter(
            (_, param_index) => masks[index][param_index] === '0',
          )
        : [],
      (used_parameters, return_parameters) => {
        const return_mask = parameters
          .map(
            parameter =>
              return_parameters.indexOf(parameter) === -1 ? '1' : '0',
          )
          .join('');
        const return_type_declaration =
          type_declarations[reverse_masks[return_mask]];
        const return_type_generics = return_type_declaration.generics!.map(
          generic => dts.create_general_type({ name: generic.name }),
        );
        const used_parameters_generics = sort_generics(
          R.uniq(
            used_parameters.reduce(
              (current, parameter) => [
                ...current,
                ...(parameter_index =>
                  parameter_index === -1
                    ? []
                    : parameters_generics[parameter_index])(
                  parameters.indexOf(parameter),
                ),
              ],
              return_type_declaration.generics!,
            ),
          ),
        ).filter(generic => type_declaration.generics!.indexOf(generic) === -1);
        members.push(
          dts.create_object_member({
            owned: dts.create_function_declaration({
              name: undefined,
              type: dts.create_function_type({
                generics: used_parameters_generics,
                parameters: used_parameters,
                return: get_overload_return_type(),
              }),
            }),
          }),
        );

        function get_overload_return_type() {
          if (return_mask.split('').every(R.equals('1'))) {
            if (type_predicate_parameter !== null) {
              if (used_parameters.indexOf(type_predicate_parameter) !== -1) {
                return return_type; // type_predicate: value is Type (matched parameter)
              }
              if (inline_return_type) {
                return dts.boolean_type; // type_predicate: boolean (unmatched parameter)
              }
            } else if (inline_return_type) {
              return return_type;
            }
          }
          return dts.create_general_type({
            name: return_type_declaration.name,
            generics: return_type_generics,
          });
        }
      },
    );

    if (!selectable || members.length <= 1) {
      return;
    }

    members.splice(
      -1,
      0,
      ...create_selectable_signatures(members, {
        is_placeholder,
        get_selectable_kind_name,
        get_selectable_selector_name,
      }),
    );
  });

  return type_declarations;
}
