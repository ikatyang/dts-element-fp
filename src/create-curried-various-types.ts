import * as dts from 'dts-element';
import * as R from 'ramda';
import {
  create_curried_types,
  CreateCurriedDeclarationsOptions,
} from './create-curried-types';
import {
  get_placeholder_type_default,
  selectable_default,
} from './utils/constants';
import { create_masks } from './utils/create-masks';
import { create_selectable_signatures } from './utils/create-selectable-signatures';
import { has_same_parameter } from './utils/has-same-parameter';

/**
 * @hidden
 */
export function create_various_curried_types(
  name: string,
  types: { [kind: string]: dts.IFunctionType },
  options: CreateCurriedDeclarationsOptions = {},
): dts.ITypeDeclaration[] {
  const {
    selectable = selectable_default,
    get_placeholder_type = get_placeholder_type_default,
    get_selectable_kind_name,
    get_selectable_selector_name,
  } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const keys = Object.keys(types);

  const parameters_length = keys.reduce<number>(
    (length, key) =>
      length === -1 || length === types[key].parameters!.length
        ? types[key].parameters!.length
        : (() => {
            throw new Error(`curried function types must be the same length`);
          })(),
    -1,
  );

  const curried_types_declarations = keys.reduce<dts.ITypeDeclaration[][]>(
    (reduced, key) => [
      ...reduced,
      create_curried_types(`${name}_${key}`, types[key], {
        ...options,
        selectable: false,
      }),
    ],
    [],
  );

  const masks = create_masks(parameters_length);

  const function_types = keys.reduce<dts.IFunctionType[]>(
    (reduced, key) => [...reduced, types[key]],
    [],
  );

  const non_conflicts = R.repeat(0, parameters_length).map((_, index) => {
    const main_function_type = function_types[0];
    return function_types
      .slice(1)
      .every(current_function_type =>
        has_same_parameter(main_function_type, current_function_type, index),
      );
  });

  const conflict_declarations: dts.ITypeDeclaration[][] = [];
  const conflict_declarations_selectables: dts.IObjectMember[][][] = [];

  const non_conflict_rebases: { [name: string]: string } = {};
  const non_conflict_declarations: dts.ITypeDeclaration[] = [];
  const non_conflict_declarations_selectables: dts.IObjectMember[][] = [];

  curried_types_declarations[0].forEach((curried_type_declaration, index) => {
    const is_return_signature =
      index === curried_types_declarations[0].length - 1;
    const mask = masks[index];
    if (
      !is_return_signature &&
      mask
        .split('')
        .every((bit, bit_index) => bit === '0' || non_conflicts[bit_index])
    ) {
      const rebase_name = `${name}_${mask}`;
      non_conflict_declarations.push(
        dts.create_type_declaration({
          name: rebase_name,
          generics: curried_type_declaration.generics,
          type: dts.create_object_type({
            members: curried_types_declarations
              .map(
                curried_types_declaration =>
                  (curried_types_declaration[index].type! as dts.IObjectType)
                    .members! as dts.IObjectMember[],
              )
              .reduce((a, b) => [...a, ...b], []),
          }),
        }),
      );
      non_conflict_declarations_selectables.push(
        curried_types_declarations
          .map((curried_types_declaration, type_index) =>
            create_selectable_signatures(
              (curried_types_declaration[index].type! as dts.IObjectType)
                .members as dts.IObjectMember[],
              {
                group: keys[type_index],
                is_placeholder,
                get_selectable_kind_name,
                get_selectable_selector_name,
              },
            ),
          )
          .reduce((a, b) => [...a, ...b], []),
      );
      curried_types_declarations
        .map(x => x[index])
        .forEach(type_declaration => {
          non_conflict_rebases[type_declaration.name] = rebase_name;
        });
    } else {
      conflict_declarations.push(
        curried_types_declarations.map(
          curried_type_declarations => curried_type_declarations[index],
        ),
      );
      conflict_declarations_selectables.push(
        curried_types_declarations.map(curried_type_declarations => {
          const members = (curried_type_declarations[index]
            .type as dts.IObjectType).members;
          return members !== undefined && members.every(dts.is_object_member)
            ? create_selectable_signatures(members as dts.IObjectMember[], {
                get_selectable_kind_name,
                get_selectable_selector_name,
                is_placeholder,
              })
            : [];
        }),
      );
    }
  });

  const mixed_type_declarations_selectables = [
    ...non_conflict_declarations_selectables,
    ...conflict_declarations_selectables.reduce((a, b) => [...a, ...b], []),
  ];

  return [
    ...non_conflict_declarations,
    ...conflict_declarations.reduce((a, b) => [...a, ...b], []),
  ].map((type_declaration, index) => {
    const object_type = type_declaration.type;
    if (
      !dts.is_object_type(object_type) ||
      object_type.members!.some(member => !dts.is_object_member(member))
    ) {
      return type_declaration;
    }
    const repeated_function_types: { [name: string]: dts.IFunctionType } = {};
    const members: dts.IObjectMember[] = [];
    object_type.members!.forEach(
      (member: dts.IObjectMember, member_index, origin_members) => {
        const function_type = (member.owned as dts.IFunctionDeclaration).type!;
        const function_return_type = function_type.return! as dts.IGeneralType;
        if (function_return_type.name in non_conflict_rebases) {
          repeated_function_types[
            non_conflict_rebases[function_return_type.name]
          ] = function_type;
        } else {
          members.push(
            member,
            ...(selectable && origin_members.length > 1
              ? [mixed_type_declarations_selectables[index][member_index]]
              : []),
          );
        }
      },
    );
    const non_conflict_merged_members = Object.keys(
      repeated_function_types,
    ).map(key => {
      const function_type = repeated_function_types[key];
      const function_return_type = function_type.return as dts.IGeneralType;
      return dts.create_object_member({
        owned: dts.create_function_declaration({
          name: undefined,
          type: {
            ...function_type,
            return: dts.create_general_type({
              name: key,
              generics: function_return_type.generics,
            }),
          },
        }),
      });
    });

    members.push(...non_conflict_merged_members);
    if (selectable) {
      members.push(
        ...create_selectable_signatures(non_conflict_merged_members, {
          is_placeholder,
          get_selectable_kind_name,
          get_selectable_selector_name,
        }),
      );
    }
    object_type.members = members;

    return type_declaration;
  });
}
