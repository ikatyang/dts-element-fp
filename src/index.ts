export * from './create-curried-interfaces';
export * from './create-curried-declarations';
