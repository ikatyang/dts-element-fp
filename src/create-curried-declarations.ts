import * as dts from 'dts-element';
import {
  create_curried_types,
  CreateCurriedDeclarationsOptions,
} from './create-curried-types';
import { create_various_curried_types } from './create-curried-various-types';
import { get_placeholder_type_default } from './utils/constants';
import { sort_signatures } from './utils/sort-signatures';

export function create_curried_declarations(
  name: string,
  type: dts.IFunctionType | Record<string, dts.IFunctionType>,
  // istanbul ignore next
  options: CreateCurriedDeclarationsOptions = {},
): (dts.IVariableDeclaration | dts.ITypeDeclaration)[] {
  const type_declarations = dts.is_function_type(type)
    ? create_curried_types(name, type, options)
    : ((keys: string[]) =>
        keys.length === 1
          ? create_curried_types(name, type[keys[0]], options)
          : create_various_curried_types(name, type, options))(
        Object.keys(type),
      );

  // istanbul ignore next
  const { get_placeholder_type = get_placeholder_type_default } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const sorted_type_declarations = type_declarations.map(type_declaration => {
    const type_declaration_type = type_declaration.type;

    if (!dts.is_object_type(type_declaration_type)) {
      return type_declaration;
    }

    return {
      ...type_declaration,
      type: {
        ...type_declaration_type,
        members: sort_signatures(
          type_declaration_type.members as dts.IObjectMember[],
          { is_placeholder },
        ),
      },
    };
  });

  return [
    dts.create_variable_declaration({
      name,
      const: true,
      type: dts.create_general_type({
        name: sorted_type_declarations[0].name,
      }),
    }),
    ...sorted_type_declarations,
  ];
}
