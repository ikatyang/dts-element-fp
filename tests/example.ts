import * as dts from 'dts-element';
import * as dts_fp from '../src/index';

test('create_curried_interfaces', () => {
  const curried_interfaces = dts_fp.create_curried_interfaces(3);
  expect(dts.emit(curried_interfaces)).toMatchSnapshot();
});

test('create_curried_declarations (normal)', () => {
  const curried_declarations = dts_fp.create_curried_declarations(
    'append',
    dts.parse_type(
      '<T, U>(value: T, list: List<U>) => (T | U)[]',
    ) as dts.IFunctionType,
  );
  expect(dts.emit(curried_declarations)).toMatchSnapshot();
});

test('create_curried_declarations (various)', () => {
  const curried_declarations = dts_fp.create_curried_declarations('map', {
    list: dts.parse_type(
      '<T, U>(fn: (x: T) => U, list: List<T>) => U[]',
    ) as dts.IFunctionType,
    functor: dts.parse_type(
      '<T, U>(fn: (x: T) => U, functor: Functor<T>) => Functor<U>',
    ) as dts.IFunctionType,
  });
  expect(dts.emit(curried_declarations)).toMatchSnapshot();
});
