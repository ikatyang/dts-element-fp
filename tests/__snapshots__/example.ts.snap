// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create_curried_declarations (normal) 1`] = `
"declare const append: append_00;
type append_00 = {
    <T>(value: T): append_10<T>;
    <T, U>(value: T, list: List<U>): append_11<T, U>;
};
type append_10<T> = {
    <U>(list: List<U>): append_11<T, U>;
};
type append_01<U> = {
    <T>(value: T): append_11<T, U>;
};
type append_11<T, U> = (T | U)[];
"
`;

exports[`create_curried_declarations (various) 1`] = `
"declare const map: map_00;
type map_00 = {
    <T, U>(fn: (x: T) => U): map_10<T, U>;
    <T, U>(fn: (x: T) => U, list: List<T>): map_list_11<U>;
    <T, U>(fn: (x: T) => U, functor: Functor<T>): map_functor_11<U>;
};
type map_10<T, U> = {
    (list: List<T>): map_list_11<U>;
    (functor: Functor<T>): map_functor_11<U>;
};
type map_list_01<T> = {
    <U>(fn: (x: T) => U): map_list_11<U>;
};
type map_functor_01<T> = {
    <U>(fn: (x: T) => U): map_functor_11<U>;
};
type map_list_11<U> = U[];
type map_functor_11<U> = Functor<U>;
"
`;

exports[`create_curried_interfaces 1`] = `
"interface CurriedFunction1<T1, R> {
    (v1: T1): R;
}
interface CurriedFunction2<T1, T2, R> {
    (v1: T1): CurriedFunction1<T2, R>;
    (v1: T1, v2: T2): R;
}
interface CurriedFunction3<T1, T2, T3, R> {
    (v1: T1): CurriedFunction2<T2, T3, R>;
    (v1: T1, v2: T2): CurriedFunction1<T3, R>;
    (v1: T1, v2: T2, v3: T3): R;
}
"
`;
